### PR TITLE
Add cmake toggle switch around arbor backend

### PR DESCRIPTION
**Pull request by w-klijn**
During testing it was found out that building NEST with Music enabled caused problems you to the hard coding of MPI handling in the arbor backend.

- A toggle has been added to CMAKE: Dwith-arborbackend=ON (default=OFF)
- #ifdef has been added throughout the C++ code.
- Additionally the flag is added the the SLI internal config struct to allows queering

### CODE REVIEW

**Comment by H27CK** [comment]
This looks good to me, @jhnnsnk. I have yet to test this implementation, but the line targeted in `nestkernel/io_manager.cpp`  is exactly that which I removed to prevent unnecessary MPI calls.

File: CMakeLists.txt [hunk 1]
```diff
@@ -45,6 +45,7 @@ set( with-libneurosim OFF CACHE STRING "Request the use of libneurosim. Optional
 set( with-music OFF CACHE STRING "Request the use of MUSIC. Optionally give the directory, where MUSIC is installed. [default=OFF]" )
 set( with-sionlib OFF CACHE STRING "Request the use of SIONlib. Optionally give the directory where SIONlib is installed. [default=OFF]" )
 set( with-boost OFF CACHE STRING "Request compilation with Boost, required for C++ testsuite. Optionally give directory with Boost installation. [default=OFF]" )
+set( with-arborbackend OFF CACHE STRING "Request compilation with Arbor backend capability actived. Requires MPI,PYTHON & MPI4PY [default=OFF]" )
```
Comment (jougs) [inline CMakeLists.txt:48]:
```suggestion
set( with-recordingbackend-arbor OFF CACHE STRING "Request compilation of the recording backend for Arbor. Requires MPI, Python and mpi4py [default=OFF]" )
```

File: CMakeLists.txt [hunk 2]
```diff
@@ -163,6 +164,7 @@ nest_process_with_music()
 nest_process_with_sionlib()
 nest_process_with_mpi4py()
 nest_process_with_boost()
+nest_process_with_arbor_backend()
```
Comment (jougs) [inline CMakeLists.txt:167]:
```suggestion
nest_process_with_recordingbackend_arbor()
```

File: cmake/ConfigureSummary.cmake [hunk 3]
```diff
@@ -152,6 +152,12 @@ function( NEST_PRINT_CONFIG_SUMMARY )
     message( "Use Boost           : No" )
   endif ()
 
+  if ( HAVE_ARBORBACKEND )
+    message( "Use Arbor Backend   : Yes" )
+  else ()
+    message( "Use Arbor Backend   : No" )
```
Comment (jougs) [inline cmake/ConfigureSummary.cmake:158]:
```suggestion
  if ( HAVE_RECORDINGBACKEND_ARBOR )
    message( "Use recording backend Arbor   : Yes" )
  else ()
    message( "Use recording backend Arbor   : No" )
```

File: cmake/ProcessOptions.cmake [hunk 4]
```diff
@@ -638,3 +638,25 @@ function( NEST_PROCESS_WITH_MPI4PY )
...
+
+function( NEST_PROCESS_WITH_ARBOR_BACKEND )
+  if (with-arborbackend)
+	if (NOT HAVE_MPI)  
+	  message( FATAL_ERROR "arbor-backend needs MPI." )
+    endif ()
+	
+	if (NOT HAVE_PYTHON) 
+	  message( FATAL_ERROR "arbor-backend needs PYTHON." )
+	endif ()
```
Comment (jougs) [inline cmake/ProcessOptions.cmake:650]:
```suggestion
function( NEST_PROCESS_WITH_RECORDINGBACKEND_ARBOR )
  if (with-recordingbackend-arbor)
	if (NOT HAVE_MPI)  
	  message( FATAL_ERROR "Recording backend Arbor needs MPI." )
    endif ()
	
	if (NOT HAVE_PYTHON) 
	  message( FATAL_ERROR "Recording backend Arbor needs Python." )
	endif ()  
```

File: cmake/ProcessOptions.cmake [hunk 5]
```diff
@@ -638,3 +638,25 @@ function( NEST_PROCESS_WITH_MPI4PY )
...
+	  message( FATAL_ERROR "arbor-backend needs PYTHON." )
+	endif ()  
+	
+    include( FindPythonModule )	
+    
+	find_python_module(mpi4py)
+	if ( HAVE_MPI4PY )
+	  include_directories( "${PY_MPI4PY}/include" )
+	else ()
+	  message( FATAL_ERROR "CMake cannot find MPI4PY, needed for arbor-backend" )
```
Comment (jougs) [inline cmake/ProcessOptions.cmake:658]:
```suggestion
	  message( FATAL_ERROR "CMake cannot find mpi4py, needed for recording backend Arbor" )
```

File: cmake/ProcessOptions.cmake [hunk 6]
```diff
@@ -638,3 +638,25 @@ function( NEST_PROCESS_WITH_MPI4PY )
...
+    endif ()
+	set( HAVE_ARBORBACKEND ON PARENT_SCOPE )
```
Comment (jougs) [inline cmake/ProcessOptions.cmake:660]:
```suggestion
	set( HAVE_RECORDINGBACKEND_ARBOR ON PARENT_SCOPE )
```

File: libnestutil/config.h.in [hunk 7]
```diff
@@ -116,6 +116,9 @@
 /* MUSIC support enabled? */
 #cmakedefine HAVE_MUSIC 1
 
+/* ARBORBACKEND support enabled? */
+#cmakedefine HAVE_ARBORBACKEND 1
```
Comment (jougs) [inline libnestutil/config.h.in:120]:
```suggestion
/* Recording backend Arbor support enabled? */
#cmakedefine HAVE_RECORDINGBACKEND_ARBOR 1
```

File: nestkernel/CMakeLists.txt [hunk 8]
```diff
@@ -107,7 +107,7 @@ if ( HAVE_SIONLIB )
       )
 endif ()
 
-if ( HAVE_MPI )
+if ( HAVE_ARBORBACKEND )
```
Comment (jougs) [inline nestkernel/CMakeLists.txt:110]:
```suggestion
if ( HAVE_RECORDINGBACKEND_ARBOR )
```

File: nestkernel/io_manager.cpp [hunk 9]
```diff
@@ -42,9 +42,9 @@
 #include "recording_backend_ascii.h"
 #include "recording_backend_memory.h"
 #include "recording_backend_screen.h"
-#ifdef HAVE_MPI
+#ifdef HAVE_ARBORBACKEND
```
Comment (jougs) [inline nestkernel/io_manager.cpp:45]:
```suggestion
#ifdef HAVE_RECORDINGBACKEND_ARBOR
```

File: nestkernel/io_manager.cpp [hunk 10]
```diff
@@ -42,9 +42,9 @@
...
 #include "recording_backend_arbor.h"
-#endif
+#endif // HAVE_ARBORBACKEND
```
Comment (jougs) [inline nestkernel/io_manager.cpp:47]:
```suggestion
#endif // HAVE_RECORDINGBACKEND_ARBOR
```

File: nestkernel/io_manager.cpp [hunk 11]
```diff
@@ -309,9 +309,9 @@ IOManager::register_recording_backends_()
   recording_backends_.insert( std::make_pair( "ascii", new RecordingBackendASCII() ) );
   recording_backends_.insert( std::make_pair( "memory", new RecordingBackendMemory() ) );
   recording_backends_.insert( std::make_pair( "screen", new RecordingBackendScreen() ) );
-#ifdef HAVE_MPI
+#ifdef HAVE_ARBORBACKEND
```
Comment (jougs) [inline nestkernel/io_manager.cpp:312]:
```suggestion
#ifdef HAVE_RECORDINGBACKEND_ARBOR
```

File: nestkernel/io_manager.cpp [hunk 12]
```diff
@@ -309,9 +309,9 @@ IOManager::register_recording_backends_()
...
   recording_backends_.insert( std::make_pair( "arbor", new RecordingBackendArbor() ) );
-#endif
+#endif // HAVE_ARBORBACKEND
```
Comment (jougs) [inline nestkernel/io_manager.cpp:314]:
```suggestion
#endif // HAVE_RECORDINGBACKEND_ARBOR
```

File: nestkernel/mpiutil.h [hunk 13]
```diff
@@ -50,8 +53,10 @@ int broadcast( int local, MPI_Comm comm, int root );
 unsigned broadcast( unsigned local, MPI_Comm comm, int root );
 float broadcast( float local, MPI_Comm comm, int root );
 
+#ifdef HAVE_ARBORBACKEND
```
Comment (jougs) [inline nestkernel/mpiutil.h:56]:
```suggestion
#ifdef HAVE_RECORDINGBACKEND_ARBOR
```

File: nestkernel/mpiutil.h [hunk 14]
```diff
@@ -50,8 +53,10 @@ int broadcast( int local, MPI_Comm comm, int root );
...
 struct comm_info;
 comm_info get_comm_info( bool is_arbor, MPI_Comm comm );
+#endif // HAVE_ARBORBACKEND
```
Comment (jougs) [inline nestkernel/mpiutil.h:59]:
```suggestion
#endif // HAVE_RECORDINGBACKEND_ARBOR
```

File: nestkernel/mpiutil_impl.h [hunk 15]
```diff
@@ -22,6 +22,10 @@
 
 #pragma once
 
+// Generated includes:
+#include "config.h"
+
+ #ifdef HAVE_ARBORBACKEND
```
Comment (jougs) [inline nestkernel/mpiutil_impl.h:28]:
```suggestion
 #ifdef HAVE_RECORDINGBACKEND_ARBOR
```

File: nestkernel/mpiutil_impl.h [hunk 16]
```diff
@@ -204,3 +208,4 @@ get_comm_info( bool is_arbor, MPI_Comm comm )
 
 } // namespace shadow
 } // namespace arb
+#endif // HAVE_ARBORBACKEND
```
Comment (jougs) [inline nestkernel/mpiutil_impl.h:211]:
```suggestion
#endif // HAVE_RECORDINGBACKEND_ARBOR
```

File: nestkernel/recording_backend_arbor.cpp [hunk 17]
```diff
@@ -20,6 +20,10 @@
  *
  */
 
+// Generated includes:
+#include "config.h"
+
+#ifdef HAVE_ARBORBACKEND
```
Comment (jougs) [inline nestkernel/recording_backend_arbor.cpp:26]:
```suggestion
#ifdef HAVE_RECORDINGBACKEND_ARBOR
```

File: nestkernel/recording_backend_arbor.cpp [hunk 18]
```diff
@@ -340,3 +344,4 @@ nest::RecordingBackendArbor::set_status( const DictionaryDatum& d )
   // if we get here, temporaries contain consistent set of properties
   P_ = ptmp;
 }
+#endif // HAVE_ARBORBACKEND
```
Comment (jougs) [inline nestkernel/recording_backend_arbor.cpp:347]:
```suggestion
#endif // HAVE_RECORDINGBACKEND_ARBOR
```

File: nestkernel/recording_backend_arbor.h [hunk 19]
```diff
@@ -23,6 +23,10 @@
 #ifndef RECORDING_BACKEND_ARBOR_H
 #define RECORDING_BACKEND_ARBOR_H
 
+// Generated includes:
+#include "config.h"
+
+#ifdef HAVE_ARBORBACKEND
```
Comment (jougs) [inline nestkernel/recording_backend_arbor.h:29]:
```suggestion
#ifdef HAVE_RECORDINGBACKEND_ARBOR
```

File: nestkernel/recording_backend_arbor.h [hunk 20]
```diff
@@ -97,4 +101,6 @@ class RecordingBackendArbor : public RecordingBackend
 
 } // namespace
 
+#endif //HAVE_ARBORBACKEND
```
Comment (jougs) [inline nestkernel/recording_backend_arbor.h:104]:
```suggestion
#endif // HAVE_RECORDINGBACKEND_ARBOR
```

File: pynest/examples/arbor_cosim_example/nest_sender.py [hunk 21]
```diff
@@ -14,6 +14,11 @@
 import nest
 
 
+STATUS_DICT = nest.ll_api.sli_func("statusdict")
+if (not  STATUS_DICT["have_arborbackend"]):
+    print ("Arbor_backend not available. Exit testscript!")
```
Comment (jougs) [inline pynest/examples/arbor_cosim_example/nest_sender.py:19]:
```suggestion
if (not  STATUS_DICT["have_recordingbackend_arbor"]):
    print ("Recording backend Arbor not available. Exit testscript!")
```

File: sli/slistartup.cc [hunk 22]
```diff
@@ -241,6 +241,7 @@ SLIStartup::SLIStartup( int argc, char** argv )
   , ismpi_name( "is_mpi" )
   , have_gsl_name( "have_gsl" )
   , have_music_name( "have_music" )
+  , have_arborbackend_name( "have_arborbackend" )
```
Comment (jougs) [inline sli/slistartup.cc:244]:
```suggestion
  , have_recordingbackend_arbor_name( "have_recordingbackend_arbor" )
```

File: sli/slistartup.cc [hunk 23]
```diff
@@ -252,6 +253,7 @@ SLIStartup::SLIStartup( int argc, char** argv )
   , exitcode_skipped_no_threading_name( "skipped_no_threading" )
   , exitcode_skipped_no_gsl_name( "skipped_no_gsl" )
   , exitcode_skipped_no_music_name( "skipped_no_music" )
+  , exitcode_skipped_no_arborbackend_name( "skipped_no_arborbackend" )
```
Comment (jougs) [inline sli/slistartup.cc:256]:
```suggestion
  , exitcode_skipped_no_recordingbackend_arbor_name( "skipped_no_recordingbackend_arbor" )
```

File: sli/slistartup.cc [hunk 24]
```diff
@@ -465,6 +467,12 @@ SLIStartup::init( SLIInterpreter* i )
   statusdict->insert( have_music_name, Token( new BoolDatum( false ) ) );
 #endif
 
+#ifdef HAVE_ARBORBACKEND
+  statusdict->insert( have_arborbackend_name, Token( new BoolDatum( true ) ) );
+#else
+  statusdict->insert( have_arborbackend_name, Token( new BoolDatum( false ) ) );
```
Comment (jougs) [inline sli/slistartup.cc:473]:
```suggestion
#ifdef HAVE_RECORDINGBACKEND_ARBOR
  statusdict->insert( have_recordingbackend_arbor_name, Token( new BoolDatum( true ) ) );
#else
  statusdict->insert( have_recordingbackend_arbor_name, Token( new BoolDatum( false ) ) );
```

File: sli/slistartup.cc [hunk 25]
```diff
@@ -511,6 +519,7 @@ SLIStartup::init( SLIInterpreter* i )
   exitcodes->insert( exitcode_skipped_no_threading_name, Token( new IntegerDatum( EXITCODE_SKIPPED_NO_THREADING ) ) );
   exitcodes->insert( exitcode_skipped_no_gsl_name, Token( new IntegerDatum( EXITCODE_SKIPPED_NO_GSL ) ) );
   exitcodes->insert( exitcode_skipped_no_music_name, Token( new IntegerDatum( EXITCODE_SKIPPED_NO_MUSIC ) ) );
+  exitcodes->insert( exitcode_skipped_no_arborbackend_name, Token( new IntegerDatum( EXITCODE_SKIPPED_NO_ARBORBACKEND ) ) );
```
Comment (jougs) [inline sli/slistartup.cc:522]:
```suggestion
  exitcodes->insert( exitcode_skipped_no_recordingbackend_arbor_name, Token( new IntegerDatum( EXITCODE_SKIPPED_NO_RECORDINGBACKEND_ARBOR ) ) );
```

File: sli/slistartup.h [hunk 26]
```diff
@@ -55,6 +55,7 @@
 #define EXITCODE_SKIPPED_NO_THREADING 203
 #define EXITCODE_SKIPPED_NO_GSL 204
 #define EXITCODE_SKIPPED_NO_MUSIC 205
+#define EXITCODE_SKIPPED_NO_ARBORBACKEND 206
```
Comment (jougs) [inline sli/slistartup.h:58]:
```suggestion
#define EXITCODE_SKIPPED_NO_RECORDINGBACKEND_ARBOR 206
```

File: sli/slistartup.h [hunk 27]
```diff
@@ -107,6 +108,7 @@ class SLIStartup : public SLIModule
   Name ismpi_name;
   Name have_gsl_name;
   Name have_music_name;
+  Name have_arborbackend_name;
```
Comment (jougs) [inline sli/slistartup.h:111]:
```suggestion
  Name have_recordingbackend_arbor_name;
```

File: sli/slistartup.h [hunk 28]
```diff
@@ -119,6 +121,7 @@ class SLIStartup : public SLIModule
   Name exitcode_skipped_no_threading_name;
   Name exitcode_skipped_no_gsl_name;
   Name exitcode_skipped_no_music_name;
+  Name exitcode_skipped_no_arborbackend_name;
```
Comment (jougs) [inline sli/slistartup.h:124]:
```suggestion
  Name exitcode_skipped_no_recordingbackend_arbor_name;
```

File: testsuite/do_tests.sh.in [hunk 29]
```diff
@@ -220,6 +220,7 @@ junit_close ()
 #   ' 203 Skipped (Threading required),'\
 #   ' 204 Skipped (GSL required),'\
 #   ' 205 Skipped (MUSIC required),'
+#   ' 206 Skipped (ARBORBACKEND required),'
```
Comment (jougs) [inline testsuite/do_tests.sh.in:223]:
```suggestion
#   ' 206 Skipped (Recording backend Arbor required),'
```

File: testsuite/do_tests.sh.in [hunk 30]
```diff
@@ -438,7 +439,8 @@ CODES_SKIPPED=\
 ' 202 Skipped (build with-mpi=OFF required),'\
 ' 203 Skipped (Threading required),'\
 ' 204 Skipped (GSL required),'\
-' 205 Skipped (MUSIC required),'
+' 205 Skipped (MUSIC required),'\
+' 206 Skipped (ARBORBACKEND required),'
```
Comment (jougs) [inline testsuite/do_tests.sh.in:443]:
```suggestion
' 206 Skipped (Recording backend Arbor required),'
```

**Comment by jougs** [comment]
Also, Please fix the formatting errors that make Travis fail. Thanks!

File: sli/slistartup.cc (hunk 25 above)
Comment (gtrensch) [inline sli/slistartup.cc:522]: In addition to @jougs suggestion this fixes the clang-formatting.

```suggestion
  exitcodes->insert(
    exitcode_skipped_no_recordingbackend_arbor_name, Token( new IntegerDatum( EXITCODE_SKIPPED_NO_RECORDINGBACKEND_ARBOR ) ) );
```

File: pynest/examples/arbor_cosim_example/arbor_proxy.py [hunk 31]
```diff
@@ -31,8 +31,7 @@ def print_spike_array_d(to_print, force=False):
 
     # Assume that we received a spike array, no error checking
     print(print_prefix + "SPIKES: [", end='')
-    for spike in to_print:
-        msg =
+    for spike in to_print:
```
Comment (gtrensch) [inline pynest/examples/arbor_cosim_example/arbor_proxy.py:34]: pep8 formatting, trailing whitespaces removed.
```suggestion
    for spike in to_print:
```

File: pynest/examples/arbor_cosim_example/nest_sender.py (hunk 21 above)
Comment (gtrensch) [inline pynest/examples/arbor_cosim_example/nest_sender.py:19]: In addition to @jougs suggestion this fixes the pep8 formatting.
```suggestion
if (not STATUS_DICT["have_recordingbackend_arbor"]):
    print("Recording backend Arbor not available. Exit testscript!")
```

**Comment by gtrensch** [comment]
This PR can be closed. PR #1550 replaces it.

**Pull request closed without merging**